### PR TITLE
Sampled KWinners Implementation for GSC

### PR DESCRIPTION
**Pull request by akashvelu**
Sampled KWinners Implementation as described in RES-1952. Instead of choosing top k units, sample the units according to the softmax distribution over the units.

### CODE REVIEW

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinner_gsc_cnn.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+from torch import nn
+
+from nupic.torch.modules import Flatten, SparseWeights, SparseWeights2d
+
+from .sampled_kwinners import SampledKWinners, SampledKWinners2d
+
+
+class SampledKWinnerGSCSparseCNN(nn.Sequential):
+    """Sparse CNN model used to classify `Google Speech Commands` dataset as
+    described in `How Can We Be So Dense?`_ paper.
```
Comment (mvacaporale) [inline projects/gsc/experiments/sampled_kwinners/sampled_kwinner_gsc_cnn.py:31]: Maybe add a sentence saying it uses the sampled form of kwiners, just to distinguish this model from `GSCSparseCNN`.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinner_le_sparse_net.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            ),
+        )
+    else:
+        network.add_module("linear{}_relu".format(suffix), nn.ReLU())
+
+    if dropout > 0.0:
+        network.add_module("linear{}_dropout".format(suffix), nn.Dropout(dropout))
+
+
+class SampledKWinnerLeSparseNet(nn.Sequential):
```
Comment (mvacaporale) [inline projects/gsc/experiments/sampled_kwinners/sampled_kwinner_le_sparse_net.py:179]: Thanks for adding this. Originally, I didn't think we'd need a LeSparseNet, but it turns out this will be the model for our main baseline.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        self.percent_on_inference = percent_on * k_inference_factor
+        self.temperature = temperature
+        self.temperature_decay_rate = temperature_decay_rate
+        self.eval_temperature = eval_temperature
+        self.learning_iterations = 0
+        self.n = 0
+        self.k = 0
+        self.k_inference = 0
+
+    def _load_from_state_dict(self, *args, **kwargs):
```
Comment (mvacaporale) [inline projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:95]: I don't think there's a need to define this. Nothing is added on line 96. Unless there's something I'm missing, you can remove it.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    For example: ``m.apply(update_boost_strength)``
+
+    :param m: KWinner module
+    """
+    if isinstance(m, SampledKWinnersBase):
+        m.update_temperature()
+
+
+class SampledKWinnersBase(nn.Module, metaclass=abc.ABCMeta):
+    """Base SampledKWinners class.
```
Comment (mvacaporale) [inline projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:45]: It would be good to add a sentence about what these classes do.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+                self.percent_on,
+                self.temperature,
+                self.temperature_decay_rate,
+                self.eval_temperature
+            )
+        )
+
+    def update_temperature(self):
+        new_temp = self.temperature - self.temperature_decay_rate
+        self.temperature = max(1.0, new_temp)
```
Comment (mvacaporale) [inline projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:115]: What's the motivation for a minimum temperature? I get that's if `temperature=1`, you're effectively sampling from the softmax distribution of the unscaled activations. I was thinking there could be a benefit to going lower. As it decays further, you effectively approach a winners take-all.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+                self.n,
+                self.percent_on,
+                self.temperature,
+                self.temperature_decay_rate,
+                self.eval_temperature
+            )
+        )
+
+    def update_temperature(self):
+        new_temp = self.temperature - self.temperature_decay_rate
```
Comment (mvacaporale) [inline projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:114]: Why subtract the decay rate vs. multiply it? are the precedents in other probabilistic algorithms that you know that do it this way? Or is the another reason?

**Review by mvacaporale — CHANGES_REQUESTED** [review]
Looks great all in all. Just some questions about implementation. And a few minor changes.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 5 above)
Comment (akashvelu) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:115]: Good point, I'll fix that!

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 6 above)
Comment (akashvelu) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:114]: I think it depends on whether we want a linear or exponential decay on the temperature. Idk if there is a standard one way or another

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 3 above)
Comment (akashvelu) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:95]: is loading from a state dictionary a functionality that we don't want?

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 6 above)
Comment (akashvelu) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:114]: Changed it to exponential decay since that does seem more appropriate.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 3 above)
Comment (mvacaporale) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py:97]: It is, but I don't believe you need to define `_load_from_state_dict` to have the functionality. The parent class, `torch.nn.Module`, already has it. In your overridden definition here, you're simply calling the parent's implementation via `super()` and nothing else is being added. Hence, you can remove yours here.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 6 above)
Comment (mvacaporale) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py]: Sounds good. We used exponential decay for the boost strength, so I think that's more consistent. Nonetheless, it's an open question. I'll add it to the list of things to consider for Alex.

File: projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py (hunk 5 above)
Comment (mvacaporale) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py]: Thanks. With that change, there's no subtle bug. When the temperature becomes to low, the softmax distribution has only one non-zero. You can check for this and default to kwinners, like this
```
# Default to kwinners when sampling k is not possible. A low enough temperature
# has only one non-zero in the softmax distribution. However, Multinomial only
# can sample the non-zero probabilites.
if ((probs != 0).sum(dim=1) < k).any():
    # Use kwinners as default; at low temperatures, this is equivalent
    return F.kwinners(x, duty_cycles=None, k=k,
                      boost_strength=0.0,  # no boosting
                      break_ties=False,
                      relu=relu,
                      inplace=inplace)
```
^ And similarly for the 2d version

Comment (mvacaporale) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py]: BTW, You should include those comments in the code or modify them just so it's clear why we'd default to regular kwinners.

Comment (akashvelu) [reply projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py]: Good point I didn't realize this. I added this to the implementation.

**Comment by mvacaporale** [comment]
Looks good to go. Thanks!

**Review by mvacaporale — APPROVED** [review]
